.gitbook/3-protocol-devs/1-dep-specs/4-data-validation.md

```diff
@@ -0,0 +1,217 @@
+| unsigned_certificate.cid              | string | A content identifier of IPFS                                     |
+| unsigned_certificate.unique_id        | string | UniqueID of the oracle that validated the data                   |
+| unsigned_certificate.oracle_address   | string | Account address of the oracle that validated the data            |
+| unsigned_certificate.deal_id          | int    | Deal to whom the provider intends to provide data                |
+| unsigned_certificate.provider_address | string | Data provider's account address                                  |
+| unsigned_certificate.data_hash        | string | The value of hashing the original data with the SHA256 algorithm |
+| signature                             | string | The value of unsigned_certificate signed with OracleKey          |
+
+
+### Data validation process
+
+#### Data Decryption
+Provider's encrypted data(`encrypted_data_base64`) can only be decrypted by Oracle.
+
+TODO: Guide to decrypt data
+```
+secret_key = SHA256(ECDH(oracle_private_key, provider_public_key))
+
+encrypted_data = Base64.Decode(encrypted_data_base64)
+
+orgin_data = AES256GCM.Decrypt(secret_key, encrypted_data)
+```
+
+#### Data Validation
+After hashing the original data with SHA256, it is encoded with HEX and compared with `data_hash`.
+```
+compare(data_hash, HEX.Encode(SHA256(origin_data))
+```
+
+Verify that the `provider_address` of the original data matches the JWT auth token issuer of the request header.
+```
+compare(origin_data.provider_address,jwtToken.issuer)
+```
+
+Deal is retrieved from the Panacea using the dealID got from the URI path.
+```
+deal = getDealFromPanacea(deal_id)
+```
+
+Before verifying the data, it is checked whether the deal status is valid.
+If the Deal's status is invalid, Oracle does not perform verification work.
+```
+compare(deal.status, 'DEAL_STATUS_ACTIVE')
+```
+
+Validate original data with `data_schema` extracted from Deal. 
+We currently support JSON schema.
+```
+data_schema = deal.data_schema
+
+JSONSchema.Validate(data_schema, orgin_data)
+```
+
+#### Data re-encryption and store to IPFS
+
+If data validation is successful, the data must be re-encrypted and stored on IPFS. 
+
+This encrypted data is stored in IPFS to be delivered to the consumer.
+
+The combinedKey used to re-encrypt the data is generated as follows:
+```
+deal_id_bz = convertUint64ToByteArray(deal_id)
+combined_key = SHA256(append(oracle_private_key, deal_id_bz, data_hash))
+```
+
+You need to convert uint64 to byte array type. Below is an example of the go language.
+```go
+func convertUint64ToByteArray(v uint64) []byte {
+    b := make([]byte, 8)
+    b[0] = byte(v >> 56)
+    b[1] = byte(v >> 48)
+    b[2] = byte(v >> 40)
+    b[3] = byte(v >> 32)
+    b[4] = byte(v >> 24)
+    b[5] = byte(v >> 16)
+    b[6] = byte(v >> 8)
+    b[7] = byte(v)
+    return b
+}
+```
```

Comment:
Would be better to mention that uint64 must be encoded in big endian.

SUGGESTIONS
> To re-encrypt the data, a symmetric secret key must be used. The symmetric secret key can be derived by the following logic.
> ```
> deal_id_bz = convertUint64ToBigEndian(deal_id)
> combined_key = SHA256(append(oracle_private_key, deal_id_bz, data_hash))
> ```
> 
> You need to convert uint64 to a byte array with the big endian byte order like the example below written in the Go language.

Reply:
Then, actually you don't need the example code.